### DISCUSSION TITLE
Hot to set the "CCIncrementDecrementButtons" parameters?

### DISCUSSION THREAD

**Discussion by josehdx** [tttapa/Control-Surface#789] [Q&A] [answered]
Hello Team

wondering if someone can help or advice about what to change on my code.

I want to use the parameter "CCIncrementDecrementButtons" to be able to send messages to "gig performer" on that app, i will use the "midi learn" funticion to link 2 physical buttons on my board to controll the parameters on a Guitar VST ( see the picture in the end on the next link: https://community.gigperformer.com/t/changing-script-behavior-from-cc-steps-to-full-semitones/11510/9 )

My code is: 

#include <Control_Surface.h>
#include <MIDI_Interfaces/BluetoothMIDI_Interface.hpp>
//#include <Arduino.h>

//HairlessMIDI_Interface midi;
BluetoothMIDI_Interface midi;

// Instantiate a CCIncrementDecrementButtons object
CCIncrementDecrementButtons buttons {
  {0, 35},              // Button pins: 0 increments, 35 decrements
  MCU::V_POT_1,        // Increment/Decrement CC address
  1,                   // Multiplier
  MCU::V_POT_SELECT_1, // Reset note address
};

//CCButton button = { 0, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, };
 
void setup() {
  // Use the Mackie Control protocol for sending relative MIDI CC messages.
  RelativeCCSender::setMode(MACKIE_CONTROL_RELATIVE);

  Control_Surface.begin(); // Initialize Control Surface
}
 
void loop() {
  Control_Surface.loop();
}

I am able to see the BLE device on my windoes laptop, i can connect and select the BLE on MIDI Berry app, but when i route to Loop midi, i can not see any kind of data after pressing the buttons.

When i use "classic Bluettooh" and using another code i saw on youtube (also using Hairless app) i can send  some midi notes and see it on Gig performer, but with this specific part i want to use, i think something is missing.

Thank you

**Comment by josehdx** [comment]
Actually i am not getting any midi data on the midi berry debuger with any other example

I am sure i am using the right pins (0 and 35 according to the next link https://ae01.alicdn.com/kf/H9d7dbadcdcf64a6ba201dc996c521515a/LILYGO-TTGO-T-Display-1-14-Inch-LCD-ESP32-WiFi-i-kompatybilna-z-Bluetooth-p-yta.jpg)

And as i mentioned before, the pins/buttons are working fine with other codes

Any suggestion is appreciated

**Comment by tttapa** [comment, marked as answer]
To check if this is an issue with BLE rather than the buttons, try something like this:
```cpp
#include <Control_Surface.h>
BluetoothMIDI_Interface midi;

void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
  static uint8_t i = 0;
  static Timer<millis> timer {250};
  if (timer) {
    Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_3}, i);
    i = i == 127 ? 0 : i + 1;
  }
}
```
It keeps sending control change messages that you can check in a MIDI monitor.

I just tested the latest `main` version of Control Surface with version 2.0.4 of the [arduino-esp32](https://github.com/espressif/arduino-esp32) core, and it works correctly on both Ubuntu 20.04 and on Windows 11 using the [Korg MIDI-BLE driver.](https://www.korg.com/us/support/download/driver/1/305/2886/) I used https://www.roxxxtar.com/apps/miditool/ as a MIDI monitor.

**Reply by josehdx** [reply]
Thank you for your feedback and for the additional  code. I have Noticed 2 things. 

With my original code, i can get some notes on an android app (actually 2, "MIDI + BTLE" to connect to my board and "midi keyboard"), now, i am using a 3rd app in adnroid. it is called "serial Bluetooth terminal" is  a similar "data" debugger juts like the midi monitoring apps, but the MIDI apps works only with MIDI data, and the  "serial Bluetooth terminal"  logs everything.

What i found with the latest app, is that with my code and with  your code, i can see the data flowing but it looks very weird to me (attaching a file )

https://drive.google.com/file/d/174iNpuE_P0k1mjH3V5I1SrPf0srwjh3W/view?usp=sharing

If you have any additional comment, please let me know about my issue, but i thing is related to my Bluetooth  card,

When i use the "classic Bluettoth Version" and using hairless app, i also see a weird behavior when i press the buttons, i got multiple inputs with a single press or release.

So probably is related. 

Thank you

**Reply by tttapa** [reply]
> What i found with the latest app, is that with my code and with your code, i can see the data flowing but it looks very weird to me (attaching a file )

MIDI is not serial, so the results printed by a “serial bluetooth terminal” will be meaningless.

> When i use the "classic Bluettoth Version" and using hairless app, i also see a weird behavior when i press the buttons, i got multiple inputs with a single press or release.

What do you mean? What “inputs” are you getting?

**Reply by josehdx** [reply]
> > What i found with the latest app, is that with my code and with your code, i can see the data flowing but it looks very weird to me (attaching a file )
> 
> MIDI is not serial, so the results printed by a “serial bluetooth terminal” will be meaningless.
> ok, so  i guess it makes sense then
> > When i use the "classic Bluettoth Version" and using hairless app, i also see a weird behavior when i press the buttons, i got multiple inputs with a single press or release.
> 
> What do you mean? What “inputs” are you getting?

Well, for the classic version of bluetooth, i guess you know that are are 2 main profiles, and when i use the classinc one, i got this on my MIDI monitor (https://youtu.be/NKx7iCxIHxU)  for the video i am using another code, but i thing is not relevant as i want to go with the BLE version. 

That said, the important thing about the BT classic and the code, is the behavior, as seen in the YT ling, i get multiple messages when i press and release the button, it should be one message per action..

So, in conclussion, i think the BT in my board is damaged. unfortunally, i have only one for testing, and i dont think i can do a deeper Troubleshoot abouth the BT behavior. Unless you know more about this. Anyway, i am very thankfull for your support. 

Thanks!

**Reply by josehdx** [reply]
Hi Again, As i was suspecting, i think my board has some kind of problem. For example, it has an USB C port and can be powered only from one position, which is weird, i think it has some electrical noise and that is making lots of problems, but i might be wrong. Sometimes i can flash, some times i got some erros, BUT the weirdest thing is i have managed to partially used your library with my board,. i need to restart or plug again my board after flashing and reconnecting for example on my Android phone. There is a nother app that if you have android i fully recommend. and is "piano MIDI BLE USB" this app has the BLE pairing build in and as well a MIDI monitor. Here i was able to see exactly what MIDI Messages were being generated. so now i am again on track with coding.

I just would like to know how to add multiple buttons or pins for your "note button" example


Thank you!


`#include <Control_Surface.h>
BluetoothMIDI_Interface midi;


// Instantiate a NoteButton object
NoteButton button {
  0,                        // Push button on pin 
  {MIDI_Notes::C(4), C(3), CHANNEL_1}, // Note C4 on MIDI channel 1
};

void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}`

**Reply by tttapa** [reply]
For future reference: you can use an array, as demonstrated in https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d7d/md_pages_Getting-Started.html

**Reply by josehdx** [reply]
you are the man!

**Comment by josehdx** [comment]
I have found a similar question/answer.Thanks for the support
